src/backend/distributed/master/master_repair_shards.c

```diff
@@ -37,118 +39,192 @@
+	 */
+	ShardInterval *shardInterval = LoadShardInterval(shardId);
+	Oid distributedTableId = shardInterval->relationId;
+	List *colocatedTableList = ColocatedTableList(distributedTableId);
+	ListCell *colocatedTableCell = NULL;
+	List *colocatedShardList = ColocatedShardIntervalList(shardInterval);
+	ListCell *colocatedShardCell = NULL;
+
+	/* check that user has owner rights in all co-located tables */
+	foreach(colocatedTableCell, colocatedTableList)
 	{
-		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-						errmsg("cannot repair shard"),
-						errdetail("Repairing shards backed by foreign tables is "
-								  "not supported.")));
+		Oid colocatedTableId = lfirst_oid(colocatedTableCell);
+		char relationKind = '\0';
+		EnsureTableOwner(colocatedTableId);
+
+		relationKind = get_rel_relkind(colocatedTableId);
+		if (relationKind == RELKIND_FOREIGN_TABLE)
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot repair shard"),
+							errdetail("Repairing shards backed by foreign tables is "
+									  "not supported.")));
+		}
 	}
 
-	targetNode = palloc0(sizeof(WorkerNode));
-	targetNode->inWorkerFile = true;
-	strlcpy(targetNode->workerName, targetPlacement->nodeName, WORKER_LENGTH);
-	targetNode->workerPort = targetPlacement->nodePort;
+	/* we sort colocatedShardList so that lock operations will not cause any deadlocks */
+	colocatedShardList = SortList(colocatedShardList, CompareShardIntervalsById);
 
-	/* retrieve DDL commands needed to drop and recreate table*/
-	ddlCommandList = RecreateTableDDLCommandList(distributedTableId);
+	/*
+	 * FIXME
+	 * To reviewer, I am not sure about locking structure here;
+	 * Are these correct lock functions?
+	 * Should we take lock on relationId instead of shardId?
+	 * Should we take lock on colocationId?
+	 */
+	foreach(colocatedShardCell, colocatedShardList)
+	{
+		ShardInterval *colocatedShard = (ShardInterval *) lfirst(colocatedShardCell);
+		uint64 colocatedShardId = colocatedShard->shardId;
 
-	/* remove existing (unhealthy) placement row; CreateShardPlacements will recreate */
-	DeleteShardPlacementRow(targetPlacement->shardId, targetPlacement->nodeName,
-							targetPlacement->nodePort);
+		/*
+		 * We've stopped data modifications of this shard, but we plan to move
+		 * a placement to the healthy state, so we need to grab a shard metadata
+		 * lock (in exclusive mode) as well.
+		 */
+		LockShardDistributionMetadata(colocatedShardId, ExclusiveLock);
 
-	/* finally, drop/recreate remote table and add back row (in healthy state) */
-	CreateShardPlacements(distributedTableId, shardId, ddlCommandList, relationOwner,
-						  list_make1(targetNode), 0, 1);
+		EnsureShardCanBeMoved(colocatedShardId, sourceNodeName, sourceNodePort);
+	}
 
-	HOLD_INTERRUPTS();
+	/* CopyShardPlacement function copies given shard with its co-located shards */
+	CopyShardPlacement(shardId, sourceNodeName, sourceNodePort, targetNodeName,
```

Comment:
In future we do not drop targetPlacement but sourcePlacement. Basically,

master_move: Drops source placement after successfull move
master_copy(restricted): Drops target placement (to recreate it completely)
master_copy(unrestricted): Does not drop target placement (because there is not any)